### DISCUSSION TITLE
Riverpod Scoping for Monorepos

### DISCUSSION THREAD

**Discussion by yogi7y** [rrousselGit/riverpod#1671] [Q&A] [answered]
How do we scope providers so that each package has its own `ProviderScope` in which all of the providers defined within the package reside rather than the global `ProviderScope`, which is located at the root above the `MaterialApp` of the client app(app using the package, in our case `app_one`)?

```
.
├── apps
│   ├── app_one
│   └── app_two
├── packages
│   ├── package_one
│   └── package_two
├── .gitignore
└── melos.yaml

```

For e.g., Let's assume `package_one` has 4 screens, If we define `ProviderScope` on the first screen it is only accessible to children on that screen and not the remaining screens. We can continue to include ProviderScope at the top of every screen but is it the only or the correct way? 

Other solutions that came to my mind were:
1. Insert `MaterialApp` to the root of every package. This helps with the scoping issue but then also scopes Themes, Navigator etc. Also for some packages bottom sheet is the entry point, I am not sure how to achieve this with the MaterialApp approach. (Please feel free to link any good resources on this.)
2. Using nested navigation?

What's the best possible way to deal with this problem?
Thanks!

**Comment by rrousselGit** [comment, marked as answer]
You _need_ a ProviderScope above everything

You can introduce scoped on a per-page basis if you want, but you still need one above the entire app. It's unavoidable for now.
